pkg/bgpv1/manager/reconcilerv2/paths.go

```diff
@@ -152,34 +153,76 @@ func reconcilePaths(params *reconcilePathsParams) (PathMap, error) {
 
 	// withdraw unneeded adverts
 	for advrtKey, advrt := range toWithdraw {
-		l.WithFields(logrus.Fields{
-			types.PathLogField:   advrt.NLRI.String(),
-			types.FamilyLogField: advrt.Family.String(),
-		}).Debug("Withdrawing path")
-
-		if err := params.Router.WithdrawPath(params.Ctx, types.PathRequest{Path: advrt}); err != nil {
+		err := withdrawPath(params, runningAdverts, advrtKey, advrt)
+		if err != nil {
 			return runningAdverts, err
 		}
-		delete(runningAdverts, advrtKey)
 	}
 
 	// create new adverts
 	for advrtKey, advrt := range toAdvertise {
-		l.WithFields(logrus.Fields{
-			types.PathLogField:   advrt.NLRI.String(),
-			types.FamilyLogField: advrt.Family.String(),
-		}).Debug("Advertising path")
-
-		advrtResp, err := params.Router.AdvertisePath(params.Ctx, types.PathRequest{Path: advrt})
+		err := advertisePath(params, runningAdverts, advrtKey, advrt)
 		if err != nil {
 			return runningAdverts, err
 		}
-		runningAdverts[advrtKey] = advrtResp.Path
 	}
 
 	return runningAdverts, nil
 }
 
+func advertisePath(params *reconcilePathsParams, runningAdverts PathMap, pathKey string, path *types.Path) error {
+	if params.PathReferences != nil && params.PathReferences[pathKey] > 0 {
+		// path already advertised for another resource
+		params.PathReferences[pathKey] += 1
+		runningAdverts[pathKey] = path
+		return nil
+	}
+
+	params.Logger.WithFields(logrus.Fields{
+		types.PathLogField:   path.NLRI.String(),
+		types.FamilyLogField: path.Family.String(),
+	}).Debug("Advertising path")
+
+	advrtResp, err := params.Router.AdvertisePath(params.Ctx, types.PathRequest{Path: path})
+	if err != nil {
+		return err
+	}
+
+	// update only in case of no error
+	if params.PathReferences != nil {
+		params.PathReferences[pathKey] += 1
+	}
+	runningAdverts[pathKey] = advrtResp.Path
+
+	return nil
+}
+
+func withdrawPath(params *reconcilePathsParams, runningAdverts PathMap, pathKey string, path *types.Path) error {
+	if params.PathReferences != nil && params.PathReferences[pathKey] > 1 {
+		// path still needs to be advertised for another resource
+		params.PathReferences[pathKey] -= 1
+		delete(runningAdverts, pathKey)
+		return nil
+	}
+
+	params.Logger.WithFields(logrus.Fields{
+		types.PathLogField:   path.NLRI.String(),
+		types.FamilyLogField: path.Family.String(),
+	}).Debug("Withdrawing path")
+
+	if err := params.Router.WithdrawPath(params.Ctx, types.PathRequest{Path: path}); err != nil {
+		return err
+	}
+
+	// update only in case of no error
+	if params.PathReferences != nil {
+		params.PathReferences[pathKey] -= 1
```

Comment:
nit: Should we set it to 0, prevent going it to negative.

Reply:
not really, as we may have more than 2 services sharing the VIP (e.g. in case of 3 svcs: when one of them is deleted we want to decrement this, not set to 0)

Reply:
My understanding was that svc is deleted and path need to be removed, we will decrement ref count at line 203 and return. 

Once we reach line 219, path is removed (line 213) and if everything is accounted for, PathReference[path] will be zero at this point.

Reply:
ah indeed - sorry, replied before taking a closer look at the exact place where the comment was applied